url-shortener/frontend/src/Frontend.hs

```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE TypeApplications #-}
+module Frontend where
+
+import qualified Data.Map as M
+import Data.Text (Text)
+import Obelisk.Frontend
+import Obelisk.Route
+import Reflex.Dom.Core
+import Control.Monad.Fix (MonadFix)
+import Language.Javascript.JSaddle (MonadJSM)
+
+import Common.Api
+import Common.Route
+import Obelisk.Generated.Static
+
+type AppWidget js t m =
+  ( DomBuilder t m
+  , MonadFix m
+  , MonadHold t m
+  , TriggerEvent t m
+  , PostBuild t m
+  , PerformEvent t m
+  , Prerender js t m
+  )
+
+type FrontendWidget js t m =
+  ( AppWidget js t m
+  , MonadJSM m
+  , MonadJSM (Performable m)
+  , HasJSContext (Performable m)
+  )
+
+data State = NotStarted | Loading | Loaded (Maybe Text)
+  deriving (Eq)
+
+submitAttrs :: State -> M.Map Text Text
+submitAttrs = \case
+  Loading -> ("disabled" =: "true")
+  _ -> mempty
+
+urlInput :: AppWidget js t m => m (Dynamic t State)
+urlInput = do
+  rec
+    inputEl <- inputElement $ def
+    (submitBtn,_) <- elAttr "span" ("id" =: "submit") $
+      elDynAttr' "button" (submitAttrs <$> state) $ text "shorten"
+    let click = domEvent Click submitBtn
+    let url = tagPromptlyDyn (_inputElement_value inputEl) click
```

Comment:
Avoid "promptly" things if at all possible since they can cause cycles in the event graph.

Reply:
You should be able to use `tag (current $ _inputElement_value inputEl) click`